### PR TITLE
CSHARP-4360: Add log messages to SDAM spec

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Logging/StructuredLogTemplateProvidersSdam.cs [hunk 1]
```diff
@@ -35,18 +36,18 @@ private static void AddSdamTemplates()
...
-                (e, _) => GetParams(e.ConnectionId, "Heartbeat succeeded"));
+                SdamCommonParams(Awaited, DurationMS, Reply),
+                (e, o) => GetParams(e.ConnectionId, "Server heartbeat succeeded", e.Awaited, (long)e.Duration.TotalMilliseconds, DocumentToString(e.Reply, o)));
 
             AddTemplateProvider<ServerHeartbeatFailedEvent>(
                 LogLevel.Debug,
-                SdamCommonParams(),
-                (e, _) => GetParams(e.ConnectionId, "Heartbeat failed"));
+                SdamCommonParams(Awaited, DurationMS, Failure),
+                (e, o) => GetParams(e.ConnectionId, "Server heartbeat failed", e.Awaited, (long)e.Timestamp.Millisecond, FormatCommandException(e.Exception, o)));
```
Comment (JamesKovacs) [inline src/MongoDB.Driver.Core/Core/Logging/StructuredLogTemplateProvidersSdam.cs:50]: `e.Timestamp` returns a `DateTime`. `DateTime.Millisecond` returns the millisecond portion of the `DateTime`, which is a number between 0 and 999. I don't think this is what you're looking for.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Looks good. Minor comment on logging the wrong component of a `DateTime`. See inline for details.

File: src/MongoDB.Driver.Core/Core/Logging/StructuredLogTemplateProvidersSdam.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Logging/StructuredLogTemplateProvidersSdam.cs]: Good catch, thanks!
I had marked this as TODO, and forgot. Added `Duration` field.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**